### PR TITLE
Message persister

### PR DESCRIPTION
**Pull request by alexsteffensen**

### CODE REVIEW

**Comment by alexsteffensen** [comment]
This closes issue #85

File: server/src/managers/message_persister.rs [hunk 1]
```diff
@@ -0,0 +1,412 @@
...
+            message_cache,
+            messages_manager: message_manager,
+            account_manager,
+            db,
+        }
+    }
+
+    pub async fn start(
+        mut message_persister: MessagePersister<T, U>,
+        flag: Arc<AtomicBool>,
```
Comment (MJHC) [inline server/src/managers/message_persister.rs:54]: make this flag be owned by the persister

File: server/src/managers/message_persister.rs [hunk 2]
```diff
@@ -0,0 +1,412 @@
...
+
+    pub async fn stop(flag: Arc<AtomicBool>) {
+        flag.store(true, Ordering::Relaxed);
+    }
+
+    // Finds the message queues where the oldest message is >10 minutes old.
+    async fn persist_next_queues(&mut self) -> Result<()> {
+        let mut queues_to_persist: Vec<String> = Vec::new();
+        let time = SystemTime::now();
+        let time_in_secs: u64 = time.duration_since(UNIX_EPOCH)?.as_secs();
```
Comment (UnluckyBird) [inline server/src/managers/message_persister.rs:71]: Use expect to unwrap, as this cannot fail

File: server/src/managers/message_persister.rs [hunk 3]
```diff
@@ -0,0 +1,412 @@
...
+        let mut queues_persisted: Vec<String> = Vec::new();
+
+        while {
+            let time = SystemTime::now();
+            let time_in_secs: u64 = time.duration_since(UNIX_EPOCH)?.as_secs();
```
Comment (UnluckyBird) [inline server/src/managers/message_persister.rs:76]: Same here

**Review by MJHC — CHANGES_REQUESTED** [review]

File: server/src/managers/message_persister.rs [hunk 4]
```diff
@@ -0,0 +1,412 @@
...
+/*
+ * Takes the message queues from the cache that is >10 minutes old,
+ * removes them from the cache and puts them into the database using the MessagesManager
+*/
+impl<T, U> MessagePersister<T, U>
+where
+    T: SignalDatabase + Send + 'static + Sync,
+    U: MessageAvailabilityListener + Send + 'static,
+{
+    pub fn new(
```
Comment (UnluckyBird) [inline server/src/managers/message_persister.rs:38]: Any reason not to combine new and start?

**Review by UnluckyBird — APPROVED** [review]

**Review by MJHC — DISMISSED** [review]
output
vscod

**Review by andreasalstrup — CHANGES_REQUESTED** [review]
failures:

---- managers::message_persister::message_persister_tests::test_message_persister_new_and_late_msg stdout ----
thread 'managers::message_persister::message_persister_tests::test_message_persister_new_and_late_msg' panicked at server/src/managers/message_persister.rs:450:9:
assertion `left == right` failed
  left: 7
 right: 5
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

---- managers::message_persister::message_persister_tests::test_message_persister_new_msg stdout ----
thread 'managers::message_persister::message_persister_tests::test_message_persister_new_msg' panicked at server/src/managers/message_persister.rs:418:9:
assertion `left == right` failed
  left: 2
 right: 1


failures:
    managers::message_persister::message_persister_tests::test_message_persister_new_and_late_msg
    managers::message_persister::message_persister_tests::test_message_persister_new_msg

test result: FAILED. 62 passed; 2 failed; 10 ignored; 0 measured; 0 filtered out; finished in 8.71s

File: server/src/managers/message_persister.rs [hunk 5]
```diff
@@ -0,0 +1,453 @@
...
+            - 300;
+
+        let (
+            handle_persisted_messages_evoked,
+            no_queue_lock_keys_in_cache,
+            old_msg_deleted,
+            msg_before,
+            msg_after,
+        ) = message_persister_test_setup_run(vec![message_time]).await;
+        assert_eq!(handle_persisted_messages_evoked, false);
```
Comment (andreasalstrup) [inline server/src/managers/message_persister.rs:415]: space :)

**Review by Darkros1245 — APPROVED** [review]

**Review by andreasalstrup — APPROVED** [review]

**Review by Darkros1245 — APPROVED** [review]

**Pull request merged**